docs/design/tools/illink/feature-switch-attributes.md

```diff
@@ -0,0 +1,482 @@
+  - `VerifyOpenGenericServiceTrimmability` property
+  - `"Microsoft.Extensions.DependencyInjection.VerifyOpenGenericServiceTrimmability"`
+
+- The "feature" of supporting or not supporting globalization is represented as:
+  - `GlobalizationMode.Invariant` property
+  - `"System.Globalization.Invariant"`
+
+No warnings are produced just because these features are enabled/disabled. Instead, they are designed to be used to change the behavior, and possibly remove unneccessary code when publishing. It's easy to imagine adding an attribute like `RequiresVariantGlobalization` that would be used to annotate APIs that rely on globalization support, such that analysis warnings would prevent accidentally pulling in the globalization stack from code that is supposed to be using invariant globalization. We don't want to do this for every feature; typically we only do this for features that represent large cross-cutting concerns, and are not available with certain app models (trimming, AOT compilation, single-file publishing).
+
+However, any attribute-based model that we pick to represent features should be able to tie in with all three representations.
+
+### Unified attribute model for feature switches and guards
+
+We would like for the attribute model to have a consistent way to refer to a feature in `FeatureGuardAttribute` or `FeatureSwitchAttribute`. The proposal is to support an attribute model for features that have an associated `Requires` attribute, and use that attribute type uniformly to refer to the feature from `FeatureGuardAttribute` and `FeatureSwitchAttribute`.
+
+For example, the feature switch for "dynamic code support" might look like this:
+```csharp
+public class RuntimeFeature
+{
+    [FeatureSwitch<RequiresDynamicCode>]
+    public static bool IsDynamicCodeSupported => // ...
+}
+
+class RequiresDynamicCodeAttribute : Attribute, IFeatureAttribute<RequiresDynamicCodeAttribute> {
+    public static string FeatureName => "System.Runtime.CompilerServices.RuntimeFeature.IsDynamicCodeSupported";
+}
+```
+and a feature guard for "dynamic code compilation" might look like this:
+```csharp
+public class RuntimeFeature
+{
+    [FeatureGuard<RequiresDynamicCode>]
+    public static bool IsDynamicCodeSupported => // ...
+}
+```
+
+The attribute definitions to support this might look like this:
+```csharp
+class FeatureGuardAttribute<T> : Attribute
+  where T : Attribute, IFeatureAttribute<T> {}
+
+interface IFeatureAttribute<TSelf> where TSelf : Attribute {
+    static abstract string FeatureName { get; }
+}
+```
+
+We could use this model even for feature switches similar to `StartupHookSupport`, which don't currently have a `StartupHoopSupportAttribute`. There's no need to actually annotate related APIs with the attribute if that is overkill for a small feature. In this case the attribute definition would just serve as metadata that allows us to define a feature switch in a uniform way.
```

Comment:
On the surface this doesn't sound right to me - declaring an attribute which will do nothing when applied to something. 
On the other hand - it "might" work if we introduce the attribute as internal - in that case nobody would be able to apply it to anything anyway, but it would still serve the purpose as the ID of the feature.
Or similarly make the constructor of the attribute private -> same effect.

I'm still not sure I like this, but it would work.

One thing I do like about this - we're not using strings to identify things, we use IL metadata - which is nice.